runtime/near-vm-logic/src/logic.rs

```diff
@@ -812,6 +813,93 @@ impl<'a> VMLogic<'a> {
         self.internal_write_register(register_id, value_hash)
     }
 
+    /// Compute multiexponentiation for (base, exp), serialized at bytes in LE notation.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + alt_bn128_g1_multiexp_base + alt_bn128_g1_multiexp_byte * num_bytes + alt_bn128_g1_multiexp_element * num_elements / (ln(min(1,num_elements))+1)`
+    pub fn alt_bn128_g1_multiexp(
+        &mut self,
+        value_len: u64,
+        value_ptr: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        self.gas_counter.pay_base(alt_bn128_g1_multiexp_base)?;
+        let value_buf = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+        self.gas_counter.pay_per_byte(alt_bn128_g1_multiexp_byte, value_buf.len() as u64)?;
+
+        let value = <Vec<(G1, Fr)>>::try_from_slice(&value_buf)
+            .map_err(|e| HostError::GuestPanic { panic_msg: format!("{}", e) })?;
+
+        let gas_estimation = {
+            let l = value.len();
+            if l == 0 {
+                0
+            } else {
+                let l = l as f64;
+                (l / (l.ln() + 1.0)).ceil() as u64
+            }
+        };
```

Comment:
We shouldn't use `f64`. What if there is an overflow or division by 0, or some other math errors.

Reply:
x/(ln(x)+1) is contraction mapping for x>=1. So, no overwflows and zero divisions here.

Also, such logarithms are used in go-ethereum for the same case (parameters estimation for sublinear pippenger multiexp) [here](https://github.com/ethereum/go-ethereum/blob/master/crypto/bls12381/g2.go#L395).

I can implement an integer logarithm, but it seems better to use the common one.

Reply:
In general, floating-point functions like `ln` are not guaranteed to produce identical results on all systems. For this reason, we shouldn't use them.

Reply:
Ok. I will use a custom implementation of integer log2floor.

Could you comment on the proposal here https://github.com/nearprotocol/NEPs/issues/98?

Reply:
We may also need to update `pippenger` implementation to not use `f64`